storage/innobase/handler/ha_innodb.cc

```diff
@@ -8343,8 +8369,17 @@ ha_innobase::write_row(
 
 	innobase_srv_conc_enter_innodb(m_prebuilt);
 
+	vers_set_fields = (table->versioned() &&
+		(sql_command != SQLCOM_CREATE_TABLE || table->s->vtmd))
+		?
+		ROW_INS_VERSIONED :
+		ROW_INS_NORMAL;
+
 	/* Step-5: Execute insert graph that will result in actual insert. */
-	error = row_insert_for_mysql((byte*) record, m_prebuilt);
+	error = row_insert_for_mysql((byte*) record, m_prebuilt, vers_set_fields);
+
+	if (m_prebuilt->trx->vers_update_trt)
+		thd_vers_update_trt(m_user_thd, true);
```

Comment:
> I understand that in the current implementation, the added parameter in row_insert_for_mysql() is needed for distinguishing UPDATE and INSERT from each other. For that, we can use a bool parameter with a default value, so that the INSERT code path will not need any code change.

@kevgs Please, replace `ins_mode != ROW_INS_NORMAL` with `table->versioned()` and `ins_mode == ROW_INS_HISTORICAL` with `vers_historical` flag.

Reply:
> Do we really need to distinguish ROW_INS_VERSIONED and ROW_INS_NORMAL here?

Yes. `ALTER TABLE ... ALGORITHM=COPY` and `versioning_alter_history=keep` case. We just copy alive rows from old to new table.

Reply:
After I submitted the patch to remove `vers_update_trt` (https://github.com/tempesta-tech/mariadb/pull/343, https://github.com/tempesta-tech/mariadb/issues/339), and the clean-up to ´row_update_for_mysql()` (https://github.com/MariaDB/server/commit/9db6fd77b681ad63449add0ffb682f9222f6c659), I can live with the DML changes.

Yesterday, I tried to see how to clean up `row_insert_for_mysql()`, and I do not think that it can be improved much. @kevgs, @midenok you can decide if any clean-up is feasible.

When it comes to InnoDB, my main remaining concerns are https://github.com/tempesta-tech/mariadb/issues/315 and https://github.com/tempesta-tech/mariadb/issues/353. Both can be fixed after merging to 10.3.

Reply:
There is one more cleanup https://github.com/tempesta-tech/mariadb/pull/359 which is about to merge soon.